### ISSUE TITLE
Dapper, mysql function - delayed exception

### ISSUE THREAD

**Issue by mbiernatek** [StackExchange/Dapper#947]
Hi,

I'm trying to use Dapper to execute mysql function that returns a decimal value. Sometime function is being executed with wrong parameter - in that case database throws exception. Everything work ok in below case:

```
string selectSql = "select public.sampleFunction(12, false)";
var results = await connection.QueryAsync<double>(selectSql);
```

If parameters are correct then function returns value, if no it throws exception immediately. What I don't understand is how code behaves in below case:

```
string selectSql = "select public.sampleFunction(12, false) union all select public.sampleFunction(13, false)";
var results = await connection.QueryAsync<double>(selectSql);
```

Again if parameters are correct then everything goes smoothly, if first function invocation throws error then exception is throwed immediately. But if first invocation is ok but second invocation (the one after union all) is wrong, then exception is being throwed affter 30 seconds (command timeout). How could I make above code to throws exception immediately? I can't change mysql function to compute values in bulk so I need to call function one by one.

Thanks

**Comment by mgravell** [comment]
The first question I have is: does it behave in the same (broken) way if
you use raw ADO.NET to do the same thing, when using the async API? I'm not
really a mysql user, but if this is a bug in the ADO.NET provider, it might
be hard for dapper to compensate. However, if the problem is something that
dapper is doing: then we might be able to help.

**Comment by mbiernatek** [comment]
I tried below code with ado.net, exception is being thrown immediately if second function invocation has wrong parameter. It throws:
"Fatal error encountered during data read."

```
try
{
	string selectValuesSql = $@"select public.sampleFunction(12, false) 
								union all 
								select public.sampleFunction(13, false)";

    var command = new MySqlCommand();
    command.Connection = connection;
    command.CommandType = System.Data.CommandType.Text;
    command.CommandText = selectValuesSql;

    connection.Open();

    MySqlDataReader reader = command.ExecuteReader();

    if (reader.HasRows)
    {
        while (reader.Read())
        {
            Console.WriteLine("{0}", reader.GetDecimal(0));
        }
    }
    else
    {
        Console.WriteLine("No rows found.");
    }

    reader.Close();

}
catch(Exception exc){  }
```

**Comment by mgravell** [comment]
OK; that is curious and suggests that dapper is doing something atypical.
Is there any chance you have a minimal example that I can try to use to
repro? Obviously I don't have your data or functions...

**Comment by mbiernatek** [comment]
Hi,

It seems that issue is with ADO.NET. I was able to reproduce the issue when MySqlDataReader is wrapped with using statement.

MySql code:
```
CREATE FUNCTION `sampleFunction`(parameter BOOLEAN) RETURNS INTEGER
BEGIN
	IF parameter = TRUE THEN
		RETURN 1;
	ELSE
		SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Custom error';
	END IF;
END
```

C# code with two versions, one with MySqlDataReader wrapped with using:

```
using (var connection = new MySqlConnection(connectionString))
using (var command = connection.CreateCommand())
{
    connection.Open();
    command.CommandText = "SELECT test.sampleFunction(true) UNION ALL SELECT test.sampleFunction(false)";

    try
    {
        using (var reader = command.ExecuteReader())
        {
            // code should throw an exception on the second iteration

            while (reader.Read())
            {
                int result = reader.GetInt32(0);
            }
        }
    }
    catch (Exception ex)
    {
        // Code reaches here after a delay of 30 seconds (same as CommandTimeout)
    }
}
```

Second without using on MySqlDataReader:

```
using (var connection = new MySqlConnection(connectionString))
using (var command = connection.CreateCommand())
{
    connection.Open();
    command.CommandText = "SELECT test.sampleFunction(true) UNION ALL SELECT test.sampleFunction(false)";

    try
    {
        // IMPORTANT: notice that reader is not wrapped in using statement
        var reader = command.ExecuteReader();

        // code should throw an exception on the second iteration
        while (reader.Read())
        {
            int result = reader.GetInt32(0);
        }

        reader.Close();
    }
    catch (Exception ex)
    {
        // Code reaches here immediately
    }
}
```

**Comment by mgravell** [comment]
That is curious, however I'm not sure it is a good idea for us to drop a
using; I suspect this example should be forwarded to the mysql provider
repo.

**Comment by bgrainger** [comment]
This definitely looks like a bug in Connector/NET (MySql.Data) to me; I can't reproduce it with MySqlConnector https://github.com/mysql-net/MySqlConnector. To report it to Oracle, file a bug at https://bugs.mysql.com/.

(This issue should be closed.)

**Comment by NickCraver** [comment]
Thanks @bgrainger! Closing out to cleanup.

**Comment by bgrainger** [comment]
FWIW Oracle just fixed the underlying problem in 8.0.20 (which shipped five days ago).

https://bugs.mysql.com/bug.php?id=89159
